Watchman.Discord/Areas/Protection/Models/ServerSafeUsers.cs

```diff
@@ -1,59 +1,71 @@
     {
-        public static UsersService UsersService { get; set; }
+        public static IComponentContext ComponentContext { get; set; }
 
         public ulong ServerId { get; }
-        public HashSet<ulong> SafeUsers { get; private set; }
+        public HashSet<ulong> SafeUsers { get; }
 
-        public ServerSafeUsers(IEnumerable<Message> serverMessages, ulong serverId, int minAverageMessagesPerWeek)
+        public ServerSafeUsers(IEnumerable<Message> serverMessages, ulong serverId, int minAverageMessagesPerWeek, HashSet<ulong> trustedRolesIds)
         {
             this.ServerId = serverId;
+            var users = ComponentContext.Resolve<DiscordServersService>().GetDiscordServerAsync(serverId).Result.GetUsers().ToDictionaryAsync(x => x.Id, x => x).Result;
             this.SafeUsers = serverMessages
                 .GroupBy(x => x.Author.Id)
-                .Where(u => IsUserSafe(u, u.Key, serverId, minAverageMessagesPerWeek))
+                .Where(u => IsUserSafe(u.ToList(), users.GetValueOrDefault(u.Key), serverId, minAverageMessagesPerWeek, trustedRolesIds))
                 .Select(x => x.Key)
                 .ToHashSet();
         }
 
-        private static bool IsUserSafe(IEnumerable<Message> userMessages, ulong userId, ulong serverId, int minAverageMessagesPerWeek)
+        private static bool IsUserSafe(IReadOnlyCollection<Message> userMessages, UserContext user, ulong serverId, int minAverageMessagesPerWeek, HashSet<ulong> trustedRolesIds)
         {
-            var days = GetHowManyDaysUserIsOnThisServer(userId, serverId);
+            if (user == null || userMessages.Count < minAverageMessagesPerWeek)
+            {
+                return false;
+            }
+            var days = GetHowManyDaysUserIsOnThisServer(user.Id, serverId);
             if (days < 30)
             {
                 return false;
             }
+            if (user.Roles.Any(x => trustedRolesIds.Contains(x.Id)))
+            {
+                return true;
+            }
 
             var weeks = userMessages.GroupBy(x => StartOfWeek(x.SentAt))
                 .Select(x => x.Count())
                 .ToList();
+
             var sum = weeks.Sum();
             var cutWeeks = weeks.Select(x =>
             {
                 var maxCountPerWeek = sum / 3;
                 return x < maxCountPerWeek ? x : maxCountPerWeek;
             });
 
-            var avg = cutWeeks.Sum() / days;
+            var avg = cutWeeks.Sum() / (days / 7);
             return avg > minAverageMessagesPerWeek;
         }
 
         private static int GetHowManyDaysUserIsOnThisServer(ulong userId, ulong serverId)
         {
-            var joinedAt = UsersService.GetUserJoinedDateTime(userId, serverId) ?? DateTime.Now;
-            return (int) (DateTime.Now.Date - joinedAt.Date).TotalDays;
+            var joinedAt = ComponentContext.Resolve<UsersService>().GetUserJoinedServerAt(userId, serverId) ?? DateTime.Now;
```

Comment:
czemu tak?

Reply:
w sensie?
tak myślę czy to Resolve nie zabiera sporo czasu jak jest wywoływane dla każdego usera
może lepiej to przypisać jednorazowo do private static field?